src/hostcalls.rs

```diff
@@ -651,6 +651,87 @@ pub fn dispatch_http_call(
     }
 }
 
+extern "C" {
+    fn proxy_grpc_call(
+        upstream_data: *const u8,
+        upstream_size: usize,
+        service_name_data: *const u8,
+        service_name_size: usize,
+        method_name_data: *const u8,
+        method_name_size: usize,
+        initial_metadata_data: *const u8,
+        initial_metadata_size: usize,
+        message_data_data: *const u8,
+        message_data_size: usize,
+        timeout: u32,
+        return_callout_id: *mut u32,
+    ) -> Status;
+}
+
+pub fn dispatch_grpc_call(
+    upstream_name: &str,
+    service_name: &str,
+    method_name: &str,
+    initial_metadata: Vec<(&str, &[u8])>,
+    message: Option<&[u8]>,
+    timeout: Duration,
+) -> Result<u32, Status> {
+    let mut return_callout_id = 0;
+    let serialized_initial_metadata = utils::serialize_bytes_value_map(initial_metadata);
+    unsafe {
+        match proxy_grpc_call(
+            upstream_name.as_ptr(),
+            upstream_name.len(),
+            service_name.as_ptr(),
+            service_name.len(),
+            method_name.as_ptr(),
+            method_name.len(),
+            serialized_initial_metadata.as_ptr(),
+            serialized_initial_metadata.len(),
+            message.map_or(null(), |message| message.as_ptr()),
+            message.map_or(0, |message| message.len()),
+            timeout.as_millis() as u32,
+            &mut return_callout_id,
+        ) {
+            Status::Ok => {
+                dispatcher::register_grpc_callout(return_callout_id);
+                Ok(return_callout_id)
+            }
+            Status::ParseFailure => Err(Status::ParseFailure),
+            Status::InternalFailure => Err(Status::InternalFailure),
+            status => panic!("unexpected status: {}", status as u32),
+        }
+    }
+}
+
+extern "C" {
+    fn proxy_grpc_cancel(token_id: u32) -> Status;
+}
+
+pub fn grpc_cancel(token_id: u32) -> Result<(), Status> {
+    unsafe {
+        match proxy_grpc_cancel(token_id) {
+            Status::Ok => Ok(()),
+            Status::NotFound => Err(Status::NotFound),
+            status => panic!("unexpected status: {}", status as u32),
+        }
+    }
+}
+
+extern "C" {
+    fn proxy_grpc_close(token_id: u32) -> Status;
+}
+
+pub fn grpc_close(token_id: u32) -> Result<(), Status> {
+    unsafe {
+        match proxy_grpc_close(token_id) {
+            Status::Ok => Ok(()),
+            Status::NotFound => Err(Status::NotFound),
+            status => panic!("unexpected status: {}", status as u32),
+        }
+    }
+}
```

Comment:
I agree with it. But host provides a path to reset gRPC call so that we should provide it on SDK side, I think.

Reply:
What I'm saying is that both `grpc_cancel` and `grpc_close` are doing exactly the same thing in Envoy for gRPC calls:

`Context::grpcClose`:
https://github.com/envoyproxy/envoy/blob/87d5eb7d44e4b24b2d098ce073f8742610c5d03e/source/extensions/common/wasm/context.cc#L1886-L1895

`Context::grpcCancel`:
https://github.com/envoyproxy/envoy/blob/87d5eb7d44e4b24b2d098ce073f8742610c5d03e/source/extensions/common/wasm/context.cc#L1913-L1922

It would be actually interesting to see what happens on the wire, i.e.is it `RST_STREAM(CANCEL)`?

Reply:
Ok. so we should provide `cancel_grpc_call` on unary call, and `cancel_grpc_stream` which send `RST_STREAM` and `close_grpc_stream` which transit the caller state to `Half-Closed`, Right?

Reply:
Right.